ludwig/utils/triton_utils.py

```diff
@@ -50,142 +98,620 @@ def forward(self, {input_signature}):
+    :param data_example: (pd.DataFrame) an example from the dataset.
+        Used to get dimensions throughout the pipeline.
     :param output_path: (str) The output path for the model repository.
     :param model_name: (str) The optional model name.
-    :param model_name: (Union[int,str]) The optional model verison.
+    :param model_version: (Union[int,str]) The optional model verison.
+    :param preprocessor_num_instances: (int) number of instances for the preprocessor (on CPU).
+    :param predictor_device_type: (str) device type for the predictor to be deployed on. One of "cpu" or "cuda"
+    :param predictor_num_instances: (int) number of instances for the predictor.
+    :param postprocessor_num_instances: (int) number of instances for the postprocessor (on CPU).
+    :param predictor_max_batch_size: (int) max_batch_size parameter for the predictor Triton config.
+    :param max_queue_delay_microseconds: (int) max_queue_delay_microseconds for all Triton configs.
 
     # Return
-    :return: (str, str) The saved model path, and config path.
+    :return: (Dict[str, Tuple[str, str, int]]) The saved config path, model path, and model size for each of
+        the three inference modules.
     """
-    model_ts = generate_triton_torchscript(model)
-    model_dir = os.path.join(output_path, model_name, str(model_version))
-    os.makedirs(model_dir, exist_ok=True)
-    # Save the file to <model_repository>/<model_name>/<model_version>/model.pt
-    model_path = os.path.join(model_dir, "model.pt")
-    model_ts.save(model_path)
-    # Save the default onfig to <model_repository>/<model_name>/config.pbtxt
-    config_path = os.path.join(output_path, model_name, "config.pbtxt")
-    with open_file(config_path, "w") as f:
-        f.write(_get_model_config(model))
-    return model_path, config_path
+
+    device_types, instance_counts = get_device_types_and_counts(
+        preprocessor_num_instances, predictor_device_type, predictor_num_instances, postprocessor_num_instances
+    )
+    split_modules = get_inference_modules(model, device_types[1])
+    example_input = get_example_input(model, device_types, data_example)
+
+    paths = {}
+    triton_masters = []
+    for i, module in enumerate(split_modules):
+        example_input = place_on_device(example_input, device_types[i])
+        triton_master = TritonMaster(
+            module,
+            example_input,
+            INFERENCE_STAGES[i],
+            predictor_max_batch_size,
+            max_queue_delay_microseconds,
+            model_name,
+            output_path,
+            model_version,
+            model.config,
+            device_types[i],
+            model_instance_count=instance_counts[i],
+        )
+        example_input = triton_master.output_data_example
+        config_path = triton_master.save_config()
+        model_path, size = triton_master.save_model()
+        paths[INFERENCE_STAGES[i]] = (config_path, model_path, size)
```

Comment:
Instead of this being a dictionary of tuples, could we simply return a list of list which include the following properties on a dictionary (or TritonArtifact class if you wanted to make it strongly typed):
* `model_name`
* `model_version`
* `platform`
* `path`
* `content_type`
* `content_length`

The rational being that in future we may need to include some additional python files or other assets which don't fall neatly into `config` or `model` categorisation.  We can also uses the `content_type` and `content_length` properties when upload to the s3 object store.

So an example output would be:
```
[
  { 
  "model_name": "titanic",
  "model_version": 1"
  "platform": "ensemble",
  "path": "titanic/config.pbtxt"
  "content_type": "text/x-protobuf",
  "content_length": 10000
  },
  { 
  "model_name": "titanic",
  "model_version": 1"
  "platform": "ensemble",
  "path": "titanic/1/placeholder.txt",
  "content_type": "text/plain",
  "content_length": 10
  },
  { 
  "model_name": "titanic_predictor",
  "model_version": 1"
  "platform": "pytorch_libtorch",
  "path": "titanic_predictor/config.pbtxt"
  "content_type": "text/x-protobuf",
  "content_length": 10
  },
  { 
  "model_name": "titanic_predictor",
  "model_version": 1"
  "platform": "pytorch_libtorch",
  "path": "titanic_predictor/1/model.pt"
  "content_type": "application/octet-stream",
  "content_length": 10
  }
]
```

Reply:
If it's a `List[List[TritonArtifact]]`, would the outer list (currently) be of size 4 (preprocessor, predictor, postprocessor, ensemble)? And the inner list would be of any size that could contain any number of `TritonArtifacts`?

In this case, if someone wanted to get a Triton config or model, would they have to iterate over the list and check the fields?

Reply:
I'm proposing it would be a flat List[TritonArtifact] where each of the stages would emit two items to the list.   The downstream implementation is just to copy these files along with their content type into a cloud bucket, to it doesn't really care that there is a series of integration stages for each model.

Yes, we would filter on the model name / version / content type if we wanted to get specific config.